host/src/FunctionsNetHost/Grpc/IncomingGrpcMessageHandler.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+
+namespace FunctionsNetHost.Grpc
+{
+    internal sealed class IncomingGrpcMessageHandler
+    {
+        private bool _specializationDone;
+        private readonly AppLoader _appLoader;
+
+        internal IncomingGrpcMessageHandler(AppLoader appLoader)
+        {
+            _appLoader = appLoader;
+        }
+
+        internal Task ProcessMessageAsync(StreamingMessage message)
+        {
+            Task.Run(() => Process(message));
+
+            return Task.CompletedTask;
+        }
+
+        private async Task Process(StreamingMessage msg)
+        {
+            if (_specializationDone)
+            {
+                // Specialization done. So forward all messages to customer payload.
+                await MessageChannel.Instance.SendInboundAsync(msg);
+                return;
+            }
+
+            var responseMessage = new StreamingMessage();
+
+            switch (msg.ContentCase)
+            {
+                case StreamingMessage.ContentOneofCase.WorkerInitRequest:
```

Comment:
Question for understanding - how do these cases compare to the `GrpcWorker` in the grpc package? Would this need to match for parity?

Reply:
Not 100%. When running in placeholder mode. this method/code path is invoked in placeholder mode. That time we do not have function metadata or capabilities (because customer code is not available yet). So we return the minimal response without them. This is by design.